main/main.cpp

```diff
@@ -1993,6 +1994,21 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 			window_vsync_mode = DisplayServer::VSyncMode::VSYNC_DISABLED;
 		}
 	}
+	{
+		frame_buffering_count = uint8_t(GLOBAL_DEF_BASIC("display/window/vsync/buffer_count", 2u));
+
+		// Triple buffering makes the best sense on the following circumstances:
+		//	1. VSync is on.
+		//	2. CPU or GPU is struggling to reach 60 FPS (assuming the monitor refreshes at 60hz)
+		//	   overall render time is between 40-75 FPS, which means sometimes the GPU has spare
+		//	   time while at other times it lacks enough time.
+		//
+		// Alternatively, it makes a lot of sense if set to VSYNC_ADAPTIVE or VSYNC_MAILBOX.
+		// When VSync off, triple buffering increases lag for no (or barely?) improvement and wastes VRAM.
+		// so we force double buffering on that case.
+		if (window_vsync_mode == DisplayServer::VSyncMode::VSYNC_DISABLED)
+			frame_buffering_count = MIN(uint8_t(2u), frame_buffering_count);
+	}
```

Comment:
This behavior should be documented in the class reference – I had to read the source code to find out about it :slightly_smiling_face:

Reply:
> Single-buffering is apparently OK with V-Sync disabled if I'm reading the code right too. This makes sense in my experience with various games.

Single buffering is almost never OK as the CPU will never overlap with the GPU to run in parallel (it can be ok in very specific non-gaming scenarios where latency is very strict and the app has control on the HW it is being run on and the performance characteristics; otherwise it's almost always a bad idea).

Anyway I'll have to edit this PR to separate buffer count from swapchain count because the simulator results are really unforgiving tying 3 swapchains to buffer count = 2 adds a lot of lag.

Reply:
> Single buffering is almost never OK as the CPU will never overlap with the GPU to run in parallel

By "OK", I meant "technically valid", rather than "advised" :slightly_smiling_face:

Reply:
Ah yes!

It is valid. That's why I forced the value to be in range [1; MAX_FRAME_LAG] (the value of MAX_FRAME_LAG is arbitrary but anything larger than that is probably useless).